### PR TITLE
12544 - Hot-seat support - ability to specify default side

### PR DESCRIPTION
**Pull request by riverwanderer**
First effort at providing a hotseat over-ride, addressing issue #12544.

If property _VassalNextSide_ is defined, that will be used.

### CODE REVIEW

**Comment by riverwanderer** [comment]
Bug to fix (occurs when VassalNextSide is not defined): 
023-09-04 17:29:31,915 [14777-AWT-EventQueue-0] ERROR VASSAL.tools.ErrorDialog - 
java.lang.NullPointerException: Cannot invoke "String.isEmpty()" because "nextChoice" is null
	at VASSAL.build.module.PlayerRoster.promptForSide(PlayerRoster.java:780)

Also, on second thoughts, it will not be sensible to set this property locally, as then any given module that uses the feature will not be able to provide a hotseat function that works for both solo and multiplayer modes.

**Comment by riverwanderer** [comment]
Retested successfully with standard hotseat side-switching and with the VassalNextSide over-ride feature.
Vassal Reference Manual update to follow.

**Review by Cattlesquat — CHANGES_REQUESTED** [review]
I see some issues with expecting a translated string in a Global Property (see below)

File: vassal-doc/src/main/readme-referencemanual/ReferenceManual/GameModule.adoc [hunk 1]
```diff
@@ -116,7 +116,11 @@ Simply type a name for each side and refer to that name in the restricted compon
 Only one player may be assigned to a side.
 When joining a game, players will be prompted to take one of the remaining available sides.
 Any number of observers (players who belong to no side) are allowed.
+
 The <<Toolbar.adoc#Retire,Retire>> or <<Toolbar.adoc#SwitchSides,Switch Sides>> button, in the main controls toolbar, allows a player to relinquish their side (making it available to the next player joining the game). You can specify the text, icon, and mouse-over tooltip for the toolbar button.
+
+The Switch Sides component includes hot-seat support. If the Module <<GlobalProperties.adoc#top,Global Property>> _VassalNextSide_ is defined and set to an available side, that side will be offered. Failing this, Vassal will determine the next side in order, excluding non-player sides. "Non-player" sides are "Solo" and those beginning "Solitaire", "Solo:", "Moderator" or "Referee". Additional sides can be listed, comma-separated, in the <<GlobalTranslatableMessages.adoc#top, Global Translatable Messages>> property _VassalSkipSideList_.
```
[the remark this answers is not in the document]
Comment (riverwanderer) [reply vassal-doc/src/main/readme-referencemanual/ReferenceManual/GameModule.adoc:122]: Yes, VassalSkipSideList is a property from PR #12636. I've removed it from there whilst I re-think this PR (12646) and #12652.

File: vassal-app/src/main/java/VASSAL/build/module/PlayerRoster.java [hunk 2]
```diff
@@ -770,19 +771,28 @@ protected String promptForSide() {
...
-    for (int tries = 0; i != myidx && tries < sides.size(); i = (i + 1) % sides.size(), tries++) { // Wrap-around search of sides
-      final String s = sides.get(i);
-      if (!alreadyTaken.contains(s) && !isSoloSide(s)) {
-        found = true; // Found an available slot that's not our current one and not a "solo" slot.
-        break;
+
+    // Automated method: set VassalNextSide (may be a Global Option ie private or a Module Global Property)
+    // If not found / available method 2 is used to find likely next side
+
+    // Reserved property VassalNextSide may override hotseat default; must be an available side
```
[the remark this answers is not in the document]
Comment (riverwanderer) [reply vassal-app/src/main/java/VASSAL/build/module/PlayerRoster.java:778]: I intended the module to work with nextSide in english only.  Re-checking the code, the **sides** array is in english right, which would be ok ?

File: vassal-app/src/main/java/VASSAL/build/module/PlayerRoster.java [hunk 3]
```diff
@@ -770,19 +771,28 @@ protected String promptForSide() {
...
-    final String mySide = getMySide(); // Get our own side, so we can find the "next" one
-    final int myidx = (mySide != null) ? sides.indexOf(mySide) : -1; // See if we have a current non-observe side.
-    int i = (myidx >= 0) ? ((myidx + 1) % sides.size()) : 0;   // If we do, start looking in the "next" slot, otherwise start at beginning.
-    for (int tries = 0; i != myidx && tries < sides.size(); i = (i + 1) % sides.size(), tries++) { // Wrap-around search of sides
-      final String s = sides.get(i);
-      if (!alreadyTaken.contains(s) && !isSoloSide(s)) {
-        found = true; // Found an available slot that's not our current one and not a "solo" slot.
-        break;
+
+    // Automated method: set VassalNextSide (may be a Global Option ie private or a Module Global Property)
```
[the remark this answers is not in the document]
Comment (riverwanderer) [reply vassal-app/src/main/java/VASSAL/build/module/PlayerRoster.java:775]: I was referring to Global Option properties being local to the user. However, I'll remove that comment as I think I already removed the suggestion from the user doc update. I don't think it's helpful in the case of nextSide.

File: vassal-app/src/main/java/VASSAL/build/module/PlayerRoster.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (riverwanderer) [reply vassal-app/src/main/java/VASSAL/build/module/PlayerRoster.java]: hmmm, yes looks like I do need to translate VassalNextSide, @Cattlesquat. I've done that in the latest PR. Will re-test.

**Comment by riverwanderer** [comment]
Tested OK.  Using a modified version of Liberté - the end of turn "solo" procedure is able to correctly default to the next side, with english and when Definition of Sides is translated to French.

**Comment by riverwanderer** [comment]
This PR doesn't create any v3.7 conflicts, can it be re-classified for the 3.7.1 milestone?

**Comment by riverwanderer** [comment]
Merged and re-tested OK. Reference Manual update included. 

Note that this change should have no effect on modules unless the designer opts-in by defining the controlling global property VassalNextSide.

**Comment by riverwanderer** [comment]
PR superseded by #12732.

**Pull request closed without merging**